### PR TITLE
Added a feature to defer non-primitive symbol lookup until runtime

### PR DESCRIPTION
**Pull request by kjmcqueen**
This is controlled by the --defer-lookup command line switch.

https://rchain.atlassian.net/browse/ROS-368
https://rchain.atlassian.net/browse/ROS-381

### CODE REVIEW

File: rosette/src/Vm.cc [hunk 1]
```diff
@@ -693,6 +693,24 @@ void VirtualMachine::evaluate(pOb expr) {
 }
 
 
+// This routine is a quick and dirty global environment lookup.  I think there
+// may be a more efficient way to do this, but I haven't researched it yet.
+int idxGlobalEnv(pOb key) {
+    int size=TAGVAL(GlobalEnv->keyVec->indexedSize());
+    const char *keyStr = BASE(key)->asCstring();
+
+    for(int i=0; i<size; i++) {
```
Comment (nashef) [inline rosette/src/Vm.cc]: Spaces around `=` and `<`, please.

File: rosette/src/Vm.cc [hunk 2]
```diff
@@ -693,6 +693,24 @@ void VirtualMachine::evaluate(pOb expr) {
 }
 
 
+// This routine is a quick and dirty global environment lookup.  I think there
+// may be a more efficient way to do this, but I haven't researched it yet.
+int idxGlobalEnv(pOb key) {
+    int size=TAGVAL(GlobalEnv->keyVec->indexedSize());
```
Comment (nashef) [inline rosette/src/Vm.cc]: Spaces around `=`.

File: rosette/src/BigBang.cc [hunk 3]
```diff
@@ -544,6 +544,7 @@ std::tuple<int, bool> BigBang(int argc, char** argv, char** envp) {
         Define("envp", GetEnvp(envp));
         Define("flag-enable-repl", GetReplFlag());
         Define("flag-verbose", VerboseFlag ? RBLTRUE : RBLFALSE);
+        Define("flag-deferLookup", DeferLookupFlag ? RBLTRUE: RBLFALSE);
```
Comment (nashef) [inline rosette/src/BigBang.cc]: I think to maintain Lisp style, it should be `flag-defer-lookup`.

File: rosette/src/CommandLine.cc [hunk 4]
```diff
@@ -105,6 +107,7 @@ int ParseCommandLine(int argc, char** argv) {
         /* Flags */
         {"verbose", no_argument, NULL, 'v'},
         {"quiet", no_argument, NULL, 'q'},
+        {"defer-lookup", no_argument, NULL, 'l'},
```
Comment (nashef) [inline rosette/src/CommandLine.cc]: In the end, what we want is the command-line interface to have a sort of `--compile` flag that just compiles and writes bytecode to the output file (possibly stdout). This doesn't look like it does that, so what's actually happening when you call it with `--defer-lookup`?

Comment (kjmcqueen) [reply rosette/src/CommandLine.cc]: I discussed this with @pyrocto yesterday.  And that is indeed the plan. To do that, we will need to create something that can serialize to disk the code and data, and be able to read a pre-compiled segment back in and execute it, whether it originates with Rosette or Rhoscala. This is just a step towards that end goal.

File: rosette/src/Vm.cc (hunk 1 above)
Comment (kjmcqueen) [reply rosette/src/Vm.cc]: Done

File: rosette/src/Vm.cc (hunk 2 above)
Comment (kjmcqueen) [reply rosette/src/Vm.cc]: Done

File: rosette/src/BigBang.cc (hunk 3 above)
Comment (kjmcqueen) [reply rosette/src/BigBang.cc]: Done

**Pull request merged**